Compilation failure on sycl::handler::fill with zero-dim accessor

**Describe the bug**
<html>
<body>
<!--StartFragment--><p style="outline: none !important; margin: 10px 0px 0px; padding: 0px; color: rgb(23, 43, 77); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Compilation failure on<span> </span><em style="outline: none !important;">sycl::handler::fill</em><span> </span>with zero-dim accessor, but there is no restrictions for<span> </span><em style="outline: none !important;">handler.fill</em><span> </span>with zero-dim accessor in SYCL2020 specification<span> </span><a href="https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:explicitmemory" class="external-link" target="_blank" rel="nofollow noopener" title="Follow link" style="outline: unset; color: rgb(0, 101, 255); text-decoration: var(--aui-link-decoration); cursor: pointer; outline-offset: unset;">4.9.4.3. SYCL functions for explicit memory operations</a>:<br style="outline: none !important;"><b style="outline: none !important;">Table 135. Member functions of the handler class</b></p><div class="table-wrap" style="outline: none !important; margin: 0px; padding: 0px; color: rgb(23, 43, 77); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

Member function | Description
-- | --
template <typename T, int Dims, access_mode Mode, target Tgt,access::placeholder IsPlaceholder>void fill(accessor<T, Dims, Mode, Tgt, IsPlaceholder> dest, const T& src) | Replicates the value of src into the memory object accessed by dest.

</div><!--EndFragment-->
</body>
</html>

**To Reproduce**
```
#include <sycl.hpp>

int main() {
  using DataT = int;
  using AccT = sycl::accessor<DataT, 0, sycl::access::mode::read_write>;

  DataT data = 1;
  sycl::range<1> r(1);
  sycl::buffer<DataT,1> data_buf(&data, r);

  sycl::queue queue;
  queue.submit([&](sycl::handler &cgh) {
    AccT acc( data_buf, cgh);
    cgh.fill(acc, 0);
  }).wait_and_throw();
  return 0;
}
```
**To build**
`clang++ -fsycl zero_dim_accessor_handler_fill.cpp -o sample`

**Environment**
- OS: Debian 6.0.12-1
- Memory: 8GB
- Compiler version: DPC++ daily 2023-04-25

**Output**
```
In file included from zero_dim_accessor_handler_fill.cpp:1:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/sycl.hpp:16:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/backend.hpp:19:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/detail/backend_traits_opencl.hpp:26:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/queue.hpp:24:
/home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/handler.hpp:2415:31: error: no matching member function for call to 'get_range'
      range<Dims> Range = Dst.get_range();
                          ~~~~^~~~~~~~~
zero_dim_accessor_handler_fill.cpp:14:9: note: in instantiation of function template specialization 'sycl::handler::fill<int, 0, sycl::access::mode::read_write, sycl::access::target::global_buffer, sycl::access::placeholder::false_t, sycl::ext::oneapi::accessor_property_list<>>' requested here
    cgh.fill(acc, 0);
        ^
/home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/accessor.hpp:2040:21: note: candidate template ignored: requirement '0 > 0' was not satisfied [with Dims = 0]
  range<Dimensions> get_range() const {
                    ^
In file included from zero_dim_accessor_handler_fill.cpp:1:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/sycl.hpp:11:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/accessor.hpp:13:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/buffer.hpp:15:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/event.hpp:15:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/detail/info_desc_helpers.hpp:13:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/info/info_desc.hpp:14:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/id.hpp:11:
/home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/detail/array.hpp:22:17: error: static assertion failed due to requirement '0 >= 1': Array cannot be 0-dimensional.
  static_assert(dimensions >= 1, "Array cannot be 0-dimensional.");
                ^~~~~~~~~~~~~~~
/home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/range.hpp:24:52: note: in instantiation of template class 'sycl::detail::array<0>' requested here
template <int dimensions = 1> class range : public detail::array<dimensions> {
                                                   ^
/home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/handler.hpp:2415:19: note: in instantiation of template class 'sycl::range<0>' requested here
      range<Dims> Range = Dst.get_range();
                  ^
zero_dim_accessor_handler_fill.cpp:14:9: note: in instantiation of function template specialization 'sycl::handler::fill<int, 0, sycl::access::mode::read_write, sycl::access::target::global_buffer, sycl::access::placeholder::false_t, sycl::ext::oneapi::accessor_property_list<>>' requested here
    cgh.fill(acc, 0);
        ^
In file included from zero_dim_accessor_handler_fill.cpp:1:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/sycl.hpp:11:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/accessor.hpp:13:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/buffer.hpp:15:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/event.hpp:15:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/detail/info_desc_helpers.hpp:13:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/info/info_desc.hpp:14:
In file included from /home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/id.hpp:15:
/home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/range.hpp:25:17: error: static assertion failed due to requirement '0 >= 1': range can only be 1, 2, or 3 dimensional.
  static_assert(dimensions >= 1 && dimensions <= 3,
                ^~~~~~~~~~~~~~~
/home/aleks/dpcpp/dpcpp_compiler/bin/../include/sycl/handler.hpp:2415:19: note: in instantiation of template class 'sycl::range<0>' requested here
      range<Dims> Range = Dst.get_range();
                  ^
zero_dim_accessor_handler_fill.cpp:14:9: note: in instantiation of function template specialization 'sycl::handler::fill<int, 0, sycl::access::mode::read_write, sycl::access::target::global_buffer, sycl::access::placeholder::false_t, sycl::ext::oneapi::accessor_property_list<>>' requested here
    cgh.fill(acc, 0);
        ^
```

--- a/sycl/include/sycl/handler.hpp
+++ b/sycl/include/sycl/handler.hpp
@@ -2429,7 +2429,7 @@ public:
     static_assert(isValidTargetForExplicitOp(AccessTarget),
                   "Invalid accessor target for the fill method.");
     if constexpr (isBackendSupportedFillSize(sizeof(T)) &&
-                  (Dims == 1 || isImageOrImageArray(AccessTarget))) {
+                  (Dims <= 1 || isImageOrImageArray(AccessTarget))) {
       setType(detail::CG::Fill);
 
       detail::AccessorBaseHost *AccBase = (detail::AccessorBaseHost *)&Dst;
@@ -2442,17 +2442,16 @@ public:
       MPattern.resize(sizeof(T));
       auto PatternPtr = reinterpret_cast<T *>(MPattern.data());
       *PatternPtr = Pattern;
+    } else if constexpr (Dims == 0) {
+      // Special case for zero-dim accessors.
+      parallel_for<
+          class __fill<T, Dims, AccessMode, AccessTarget, IsPlaceholder>>(
+          range<1>(1), [=](id<1> Index) { Dst = Pattern; });
     } else {
-      if constexpr (Dims == 0) {
-        parallel_for<
-            class __fill<T, Dims, AccessMode, AccessTarget, IsPlaceholder>>(
-            range<1>(1), [=](id<1> Index) { Dst = Pattern; });
-      } else {
-        range<Dims> Range = Dst.get_range();
-        parallel_for<
-            class __fill<T, Dims, AccessMode, AccessTarget, IsPlaceholder>>(
-            Range, [=](id<Dims> Index) { Dst[Index] = Pattern; });
-      }
+      range<Dims> Range = Dst.get_range();
+      parallel_for<
+          class __fill<T, Dims, AccessMode, AccessTarget, IsPlaceholder>>(
+          Range, [=](id<Dims> Index) { Dst[Index] = Pattern; });
     }
   }
 

--- a/sycl/include/sycl/handler.hpp
+++ b/sycl/include/sycl/handler.hpp
@@ -2443,10 +2443,16 @@ public:
       auto PatternPtr = reinterpret_cast<T *>(MPattern.data());
       *PatternPtr = Pattern;
     } else {
-      range<Dims> Range = Dst.get_range();
-      parallel_for<
-          class __fill<T, Dims, AccessMode, AccessTarget, IsPlaceholder>>(
-          Range, [=](id<Dims> Index) { Dst[Index] = Pattern; });
+      if constexpr (Dims == 0) {
+        parallel_for<
+            class __fill<T, Dims, AccessMode, AccessTarget, IsPlaceholder>>(
+            range<1>(1), [=](id<1> Index) { Dst = Pattern; });
+      } else {
+        range<Dims> Range = Dst.get_range();
+        parallel_for<
+            class __fill<T, Dims, AccessMode, AccessTarget, IsPlaceholder>>(
+            Range, [=](id<Dims> Index) { Dst[Index] = Pattern; });
+      }
     }
   }
 

--- a/sycl/source/detail/memory_manager.cpp
+++ b/sycl/source/detail/memory_manager.cpp
@@ -781,7 +781,7 @@ void MemoryManager::fill(SYCLMemObjI *SYCLMemObj, void *Mem, QueueImplPtr Queue,
 
   const detail::plugin &Plugin = Queue->getPlugin();
   if (SYCLMemObj->getType() == detail::SYCLMemObjI::MemObjType::Buffer) {
-    if (Dim == 1) {
+    if (Dim <= 1) {
       Plugin.call<PiApiKind::piEnqueueMemBufferFill>(
           Queue->getHandleRef(), pi::cast<RT::PiMem>(Mem), Pattern, PatternSize,
           Offset[0] * ElementSize, Range[0] * ElementSize, DepEvents.size(),

--- a/sycl/test-e2e/Basic/fill_accessor.cpp
+++ b/sycl/test-e2e/Basic/fill_accessor.cpp
@@ -46,7 +46,30 @@ void CheckFill(queue &Q, range<Dims> Range, T Init, T Expected) {
 }
 
 template <typename T>
+void CheckFillZeroDimAccessor(queue &Q, T Init, T Expected) {
+  constexpr int Dims = 1;
+  range<1> Range(1);
+  std::vector<T> Data(Range.size(), Init);
+  {
+    buffer<T, Dims> Buffer(Data.data(), Range);
+    Q.submit([&](handler &CGH) {
+       accessor<T, 0, sycl::access::mode::write> Accessor(Buffer, CGH);
+       CGH.fill(Accessor, Expected);
+     }).wait_and_throw();
+  }
+  for (size_t I = 0; I < Range.size(); ++I) {
+    if (Data[I] != Expected) {
+      std::cout << "Unexpected value " << Data[I] << " at index " << I
+                << " after fill. Expected " << Expected << "." << std::endl;
+      ++NumErrors;
+      return;
+    }
+  }
+}
+
+template <typename T>
 void CheckFillDifferentDims(queue &Q, size_t N, T Init, T Expected) {
+  CheckFillZeroDimAccessor<T>(Q, Init, Expected);
   CheckFill<T>(Q, range<1>{N}, Init, Expected);
   CheckFill<T>(Q, range<2>{N, N}, Init, Expected);
   CheckFill<T>(Q, range<3>{N, N, N}, Init, Expected);

